src/lib/operators/join_helper/join_output_writing.cpp

```diff
@@ -0,0 +1,308 @@
+#include "join_output_writing.hpp"
+
+#include <unordered_map>
+
+#include <boost/functional/hash_fwd.hpp>
+
+#include "hyrise.hpp"
+#include "scheduler/job_task.hpp"
+#include "storage/segment_iterate.hpp"
+
+namespace {
+
+using namespace hyrise;  // NOLINT(build/namespaces)
+
+using PosLists = std::vector<std::shared_ptr<const AbstractPosList>>;
+using PosListsByColumn = std::vector<std::shared_ptr<PosLists>>;
+
+struct PosListsHasher {
+  size_t operator()(const PosLists& pos_lists) const {
+    return boost::hash_range(pos_lists.begin(), pos_lists.end());
+  }
+};
+
+/**
+ * Returns a vector where each entry with index i references a PosLists object. If entries i and j reference the same
+ * PosLists object, both columns originate from the same table. As a consequence, they can later share the same (newly
+ * created) PosLists of the join result.
+ * @param input_table
+ */
+PosListsByColumn setup_pos_list_mapping(const std::shared_ptr<const Table>& input_table) {
+  Assert(input_table->type() == TableType::References, "Function only works for reference tables");
+
+  const auto chunk_count = input_table->chunk_count();
+  const auto column_count = input_table->column_count();
+
+  std::unordered_map<PosLists, std::shared_ptr<PosLists>, PosListsHasher> shared_pos_lists_by_pos_lists;
```

Comment:
What about this:

```
  // We use a map to recognize when columns of the input table share the same PosLists (i.e., using shared pointers to
  // the same PosList). The value of this map is later used to get the actual position lists when constructing new
  // reference segments (otherwise, we would need to cast AbstractSegment to ReferenceSegment first in order to obtain
  // the PosList).
  auto shared_pos_lists_by_pos_lists = std::unordered_map<PosLists, std::shared_ptr<PosLists>, PosListsHasher>{};
  shared_pos_lists_by_pos_lists.reserve(column_count);
```

Reply:
Yeah, I like the comment